test/jsupdate-behaviour-mocha.js

```diff
@@ -112,10 +213,115 @@ describe("framework-state", function() {
+                });
+            }).catch(fail);
+        }).should.eventually.have.property('data', "Hello again");
+    });
+    it("should fire updater if upstream updater changed output state lm", function() {
+        return new Promise(function(done, fail){
+            var count = 0;
+            test.createNetwork({
+                once: jswrapper(function(input) {
+                    switch(++count) {
+                        case 1: return input;
+                        default:
+                        var outputState = this.outputState();
+                        outputState.lm = "changed";
+                        this.outputState(outputState);
+                    }
+                }),
+                sut: jswrapper(function(input) {
+                    return "Hello " + input;
+                })
+            }).then(function(network){
+                network.graph.addEdge('once', 'output', 'sut', 'input');
+                return new Promise(function(adv) {
+                    test.onOutPortData(network.processes.sut.component, 'output', adv);
+                    test.sendData(network.processes.once.component, 'input', "once");
+                }).then(function() {
+                    test.onOutPortData(network.processes.sut.component, 'output', done);
+                    test.sendData(network.processes.once.component, 'input', "again");
+                });
+            }).catch(fail);
+        }).should.eventually.have.property('data', "Hello once");
+    });
+    it("should fire updater if upstream updater did nothing, after error, after valid output", function() {
+        return new Promise(function(done, fail){
+            var count = 0;
+            test.createNetwork({
+                upstream: jswrapper(function(input) {
+                    switch(++count) {
+                        case 1: return input;
+                        case 2: throw input;
+                        case 3: return;
+                    }
+                }),
+                sut: jswrapper(function(input) {
+                    return "Hello " + input;
+                })
+            }).then(function(network){
+                network.graph.addEdge('upstream', 'output', 'sut', 'input');
+                return new Promise(function(adv) {
+                    test.onOutPortData(network.processes.sut.component, 'output', adv);
+                    test.sendData(network.processes.upstream.component, 'input', "upstream");
+                }).then(function() {
+                    return new Promise(function(adv) {
+                        test.onOutPortData(network.processes.upstream.component, 'output', adv);
+                        test.sendData(network.processes.upstream.component, 'input', "again");
+                    });
+                }).then(function() {
+                    test.onOutPortData(network.processes.sut.component, 'output', done);
```

Comment:
To clarify the desired behavior, the upsteam node ('upstream') receives input three times, causing its updater to be fired three times.  The first time its updater is fired, the updater returns a value "upstream" that is set into its state.  This state change causes the downstream node ('sut') to be notified and sut's updater is fired because it's input state changed.  The second time 'upstream' receives input upstream's updater is again fired, but this time it throws an error, which does not cause upstream's output.data to change, but it does cause upstream's output.error to be set.  As a result, the downstream node 'sut' should be notified that the status of its input changed, but sut's shouldFireUpdater should return false, because its input is no longer ready.  The third time upstream receives input its updater is again called and this time it returns undefined, signaling that its output.data did not change.  However, because its updater returned without throwing an error, its output.error flag is cleared, causing the downstream node 'sut' to be notified that its input status changed (even though its input data did not change).  This causes sut's shouldFireUpdater to be called, but it should return false because the node's input data did not change, so sut's updater should NOT be called.

This test makes me realize that there are cases when events need to be propagated downstream even if a node's updater does not fire, because the node's state becomes stale or non-stale (i.e., invalid or valid).  Events need to be propagated downstream if either the state's data has changed (as indicated by a new LM) or its status changes (error/non-error or stale/non-stale).  I suppose this means we need another flag -- output.stale -- though maybe there is an easier way to handle it.